File: benchmark/synduce/list/zero_after_one.f
Inductive BList = nil Unit | cons {Bool, BList};
Inductive CList = emp Unit | single Bool | concat {CList, CList};

const_true = true;
const_false = false;
op_and = \x: Bool. \y: Bool. and x y;
op_or = \x: Bool. \y: Bool. or x y;

cat_list = fix (
  \f: BList -> BList -> BList. \xs: BList. \ys: BList.
  match xs with
    nil _ -> ys
  | cons {hd, tl} -> cons {hd, f tl ys}
  end
);

repr = fix (
  \f: CList -> Compress BList. \xs: CList.
  match xs with
    emp _ -> nil unit
  | single a -> cons {a, nil unit}
  | concat {a, b} -> cat_list (f a) (f b)
  end
);

spec = \xs: BList. (fix (
  \f: BList -> {Bool, Bool, Bool}. \xs: BList.
  match xs with
    nil _ -> {false, false, false}
  | cons {hd, tl} ->
    let result = f tl in
    let new_seen1 = or result.1 hd in
    let new_res = or result.2 (and result.1 (not hd)) in
    let new_aux = or result.3 (not hd) in
    {new_seen1, new_res, new_aux}
  end
) xs).2;

main = \xs: CList. spec (repr xs);

File: ml-benchmark/synduce/list/zero_after_one.ml
type blist = Nil | Cons of bool * blist
type clist = Emp | Single of bool | Concat of clist * clist

let const_true = true
let const_false = false
let op_and = fun x -> fun y -> x && y
let op_or = fun x -> fun y -> x || y

let rec cat_list xs ys =
  match xs with
  | Nil -> ys
  | Cons (hd, tl) -> Cons (hd, cat_list tl ys)

let rec repr xs =
  match xs with
  | Emp -> Nil
  | Single a -> Cons (a, Nil)
  | Concat (a, b) -> cat_list (repr a) (repr b)

let spec xs =
  let rec f xs =
    match xs with
    | Nil -> (false, false, false)
    | Cons (hd, tl) ->
      let result = f tl in
      match result with
      | (r1, r2, r3) ->
        let new_seen1 = r1 || hd in
        let new_res = r2 || (r1 && (not hd)) in
        let new_aux = r3 || (not hd) in
        (new_seen1, new_res, new_aux)
  in
  match f xs with
  | (_, r2, _) -> r2

let program xs = spec (repr xs)